### DISCUSSION TITLE
[Chapter 4] - Builder Pattern과 Stream Pipe Line

### DISCUSSION THREAD

**Discussion by Irisation23** [Study-2-Modern-Java-In-Action/Study-2-Modern-Java-in-Action#44] [CHAPTER 4 스트림 소개 - Q&A]
p152 발췌

- 스트림 파이프라인의 개념은 빌더 패턴(builder pattern)과 비슷하다. 
- 호출을 연결(중간 연산)해서 설정을 만들고 해당 설정의 마지막에 `build()`를 호출(최종 연산)한다.

### Q. 그렇다면 builder pattern은 Stream이라 명칭 할 수 있을까?

**Comment by ahngilwoong** [comment]
제가 저 부분을 그냥 넘어갔는지 기억이 안나서 다시 한번 보게 되는 계기가 되었습니다.
말씀하신대로 빌더패턴과 비슷한 경향이 있는 것 같습니다.
근데 제가 생각하기에 빌더패턴은 객체를 만들때 쓰는 패턴? 이라는 인식으로 박혀있고
스트림 파이프라인은 데이터 처리에 주 목적을 가진 느낌이라 약간 지향하는 바가 다르다..라고 느낍니다..
공교롭게도 겹친 느낌이 듭니다.

**Comment by coalong** [comment]
p. 152 
` 스트림 파이프라인의 개념`은 빌더 패턴과 비슷하다. 
빌더 패턴의 호출을 연결해서 설정을 만드는 부분 == 스트림의 중간 연산 연결
빌더 패턴의 `build` 메서드로 최종 결과 도출 == 스트림의 결과를 만들 최종 연산

구성하는 부분이 유사하게 느껴지네요.

### 그렇다면 builder pattern 은 stream 이라 명칭할 수 있을까?

라는 질문을 생각해봤을때,

p. 141 
`스트림` 이란 데이터 처리 연산을 지원하도록 소스에서 추출된 `연속된 요소(Sequence of elements)`로 정의할 수 있다.
정의 3가지 
- 연속된 요소
- 소스
- 데이터 처리 연산

라는 정의를 다시 읽어보면서 빌터 패턴은 스트림 정의에 부합하지 않다고 생각했습니다.

빌터패턴의 메서드 체인과 스트림의 파이프라인이 비슷한 개념을 띄고 있는 건 함수형 프로그래밍의 영향이 있다고 생각합니다.
